File: src/_03_BehavioralPattern/_03_03_Interpreter/java/after/PostfixParser.java
package _03_BehavioralPattern._03_03_Interpreter.java.after;

import java.util.Stack;

public class PostfixParser {

  public static PostfixExpression parse(String expression) {
    Stack<PostfixExpression> stack = new Stack<>();

    for (Character c: expression.toCharArray()) {
      stack.push(createExpression(c, stack));
    }

    return stack.pop();
  }

  private static PostfixExpression createExpression(Character c, Stack<PostfixExpression> stack) {
    switch(c) {
      case '+': {
        return new PlusExpression(stack.pop(), stack.pop());
      }

      case '-': {
        PostfixExpression rhs = stack.pop();
        PostfixExpression lhs = stack.pop();
        return new MinusExpression(lhs, rhs);
      }

      default: {
        return new VariableExpression(c);
      }
    }
  }
  
}

File: src/_03_BehavioralPattern/_03_03_Interpreter/PostfixParser.js
import VariableExpression from "./VariableExpression.js";
import PlusExpression from "./PlusExpression.js";
import MinusExpression from "./MinusExpression.js";

/**
 * @typedef { import("./PostfixExpression").default } PostfixExpression
 */

export default class PostfixParser {
  /**
   * @param { string } expression
   * @returns { PostfixExpression }
   */
  static parse(expression) {
    /** @type { PostfixExpression[] } */
    const stack = [];
    
    expression.split("").forEach(char => {
      stack.push(PostfixParser._createExpression(char, stack));
    });

    return stack.pop();
  }

  /**
   * @param { string } char 
   * @param { { [key: string]: number }[] } stack 
   */
  static _createExpression(char, stack) {
    switch(char) {
      case "+": {
        return new PlusExpression(stack.pop(), stack.pop());
      }

      case "-": {
        const rhs = stack.pop();
        const lhs = stack.pop();
        return new MinusExpression(lhs, rhs);
      }

      default: {
        return new VariableExpression(char);
      }
    }
  }
}